### PR TITLE
Register

### PR DESCRIPTION
**Pull request by vargadavid8**
POST:
{
	"firstname": "Robert",	
	"lastname": "Gida",
	"email": "rg@rg.hu",
	"password": "123456"
}

### CODE REVIEW

File: LearningTool/src/main/java/main/controllers/MainController.java [hunk 1]
```diff
@@ -55,4 +56,21 @@ public HttpStatus loginUser(@RequestBody UserDTO userDTO) {
...
+            return HttpStatus.OK;
+        }catch(Exception e){
+            return HttpStatus.FORBIDDEN;
+        }
+    }
+
+    @PostMapping("/adduser")
+    public ResponseEntity<?> addUser(@RequestBody User user) {
+            userService.addUser(user);
+            return new ResponseEntity<>(user, HttpStatus.OK);
```
Comment (rdgrv) [inline LearningTool/src/main/java/main/controllers/MainController.java]: Do we need ResponseEntity here? The status will be OK either way. I think User would be enough. If you decide to change it, make sure to change the return type and delete the import as well.

File: LearningTool/src/main/java/main/models/User.java [hunk 2]
```diff
@@ -13,12 +17,23 @@
     private String firstname;
     private String lastname;
     private String email;
-    private LocalDateTime createDate;
-    private String role;
+    private LocalDateTime createDate = LocalDateTime.now();
+
+    @Column(nullable = false)
```
Comment (rdgrv) [inline LearningTool/src/main/java/main/models/User.java]: shouldn't we add this annotation to all of the fields?

File: LearningTool/src/main/java/main/repositories/RoleRepository.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package main.repositories;
+
+import main.models.Role;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+
```
Comment (rdgrv) [inline LearningTool/src/main/java/main/repositories/RoleRepository.java]: extra line

File: LearningTool/src/main/java/main/services/UserDetailsImpl.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
+    }
+
+
```
Comment (rdgrv) [inline LearningTool/src/main/java/main/services/UserDetailsImpl.java]: extra lines

File: LearningTool/src/main/java/main/services/UserServiceImpl.java [hunk 5]
```diff
@@ -26,6 +37,18 @@ public User addUser(User user) {
         return userRepository.save(user);
     }
 
+    @Transactional
+    @Override
+    public void registerUser(User user){
+        Role userRole = roleRepository.findByRole(USER_ROLE);
+        if(userRole != null){
+            user.getRoles().add(userRole);
+        }else{
```
Comment (rdgrv) [inline LearningTool/src/main/java/main/services/UserServiceImpl.java]: make sure to put spaces where they belong, or hit ctrl+alt+L for autoformatting :)

File: LearningTool/src/main/java/main/controllers/MainController.java (hunk 1 above)
Comment (dubaimartin) [reply LearningTool/src/main/java/main/controllers/MainController.java]: I would add multiple points here:
1. We should DTOs for data transfers, hence the name: Data Transfer Object. A DTO should only contain the needed fields, getters, setters, nothing more.
2. I would not handle exceptions like this. We need to write a global exception translator to handle every exception. The result is going to be much much cleaner.
3. What is adduser and register? What is the difference?
4. We should also use the @Valid annotation, and add validation to the DTOs.

File: LearningTool/src/main/java/main/models/Role.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    private Long id;
+
+    private String role;
+
+    @ManyToMany( mappedBy = "roles")
+    private Set<User> users = new HashSet<User>();
+
+    private Role(){}
+
+    public Role(String role){
```
Comment (dubaimartin) [inline LearningTool/src/main/java/main/models/Role.java]: @vargadavid8 We don't really need parameterised constructors actually. It is not a good practice, if you think about having lots of fields, then we would need a constructor for all, for every combination? That's not a good idea. Here we could also introduce chaining the setters.

File: LearningTool/src/main/java/main/models/User.java (hunk 2 above)
Comment (dubaimartin) [reply LearningTool/src/main/java/main/models/User.java]: @rdgrv It depends if that field is actually required or not.

File: LearningTool/src/main/java/main/repositories/RoleRepository.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+@Repository
+public interface RoleRepository extends CrudRepository<Role, Long> {
+
+    Role findByRole(String role);
```
Comment (dubaimartin) [inline LearningTool/src/main/java/main/repositories/RoleRepository.java]: @vargadavid8 I actually don't understand this. Am I finding a role by a role? :) That seems a little... well.. strange.

File: LearningTool/src/main/java/main/services/UserServiceImpl.java (hunk 5 above)
Comment (dubaimartin) [reply LearningTool/src/main/java/main/services/UserServiceImpl.java]: @vargadavid8 @rdgrv Please use an Optional here to a void null checks.

File: LearningTool/src/main/java/main/controllers/MainController.java (hunk 1 above)
Comment (vargadavid8) [reply LearningTool/src/main/java/main/controllers/MainController.java]: I think we don't need the addUser (it was in an earlier version of class) , I will remove it.
All of the registered users get the role "USER". If this role doesn't exist in the roles table registerUser will add that. And the users_roles table will contain the new user_id with the related role_id.

**Pull request merged**